qcodes/utils/update/qcodes_updater.py

```diff
@@ -0,0 +1,127 @@
+import os
+import subprocess
+import inspect
+import shutil
+import datetime
+import warnings
+import git
+
+import qcodes
+
+def qcodes_backup(env_name: str = 'qcodes', env_backup_name: str = 'qcodes_backup'):
+    """
+    Back up QCoDeS Anaconda Python environment.
+
+    Args:
+        env_name: Name of the environment to be updated
+        env_backup_name: Name of the back up environment to be created
+
+    Raises:
+        git.exc.GitError: If working tree is dirty
+    """
+    # QCoDeS source path
+    _source = os.sep.join(qcodes.__file__.split(os.sep)[:-2])
+    # Make sure that git working tree is clean
+    repo = git.Repo(_source)
+    if (inspect.stack()[1].function == 'update_qcodes_installation' and repo.is_dirty()):
+        raise git.exc.GitError
+    elif (inspect.stack()[1].function == 'conda_env_update' and repo.is_dirty()):
+        raise git.exc.GitError
+    elif repo.is_dirty():
+        print("QCoDeS working tree is not clean. Please commit or stash your changes, and try again.")
```

Comment:
I suggest that we just replace the 3 sections above with raising the same exception. Then you move the alternative handling of the exception into whatever calls the function. We can define our own exception to make it explicit that we only catch that exception.